reflectionserializer/manager.go

```diff
@@ -0,0 +1,795 @@
+				}
+			}
+		}
+	}
+	return restoredStruct.Interface(), nil
+}
+
+func (m *SerializationManager) deserializePointer(valueType reflect.Type, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) (interface{}, error) {
+	// if pointer value can be nil, write first byte to indicate whether it has value
+	if fieldMetadata.AllowNil {
+		nilPointer, err := buffer.ReadByte()
+		if err != nil {
+			return nil, err
+		}
+		// if pointer prefix is 0 then set value of the pointer to nil
+		if nilPointer == 0 {
+			return nil, nil
+		}
+	}
+	p := reflect.New(valueType.Elem())
+
+	// try to use built-in deserializer if the pointer type provides it
+	processed, err := m.tryBuiltInDeserializer(p, valueType, fieldMetadata, buffer)
+	if err != nil {
+		return nil, err
+	}
+	if processed {
+		return p.Interface(), nil
+	}
+
+	// deserialize as normal pointer type
+	de, err := m.DeserializeType(valueType.Elem(), fieldMetadata, buffer)
+	if err != nil {
+		return nil, err
+	}
+	p.Elem().Set(reflect.ValueOf(de))
+	return p.Interface(), nil
+}
+
+func (m *SerializationManager) deserializeSlice(valueType reflect.Type, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) (interface{}, error) {
+	// read length of the slice and validate that it matches the bounds
+	sliceLen, err := ReadLen(fieldMetadata.LengthPrefixType, buffer)
+	if err != nil {
+		return nil, err
+	}
+	err = ValidateLength(sliceLen, fieldMetadata.MinSliceLength, fieldMetadata.MaxSliceLength)
+	if err != nil {
+		return nil, err
+	}
+	restoredSlice := reflect.New(valueType).Elem()
+	if sliceLen == 0 {
+		return restoredSlice.Interface(), nil
+	}
+
+	if valueType.Elem().Kind() == reflect.Uint8 {
+		// special handling of byte slice to optimize execution
+		restored, err := buffer.ReadBytes(sliceLen)
+		if err != nil {
+			return nil, err
+		}
+		restoredSlice = reflect.ValueOf(restored)
+	} else if fieldMetadata.LexicalOrder || fieldMetadata.NoDuplicates {
+		// if lexical ordering or no duplicates is required, perform additional processing
+		readOffset := buffer.ReadOffset()
+		elements := make([]struct {
+			elem      interface{}
+			elemBytes []byte
+		}, 0)
+		elementsMap := make(map[string]customtypes.Empty)
+		for i := 0; i < sliceLen; i++ {
+			elem, err := m.DeserializeType(valueType.Elem(), fieldMetadata, buffer)
+			if err != nil {
+				return nil, err
+			}
+			bytesRead := buffer.ReadOffset() - readOffset
+			buffer.ReadSeek(-bytesRead)
+			elemBytes, err := buffer.ReadBytes(bytesRead)
+			readOffset = buffer.ReadOffset()
+
+			elemBytesString := typeutils.BytesToString(elemBytes)
+			if err != nil {
+				return nil, err
+			}
+
+			if _, seenAlready := elementsMap[elemBytesString]; fieldMetadata.NoDuplicates && seenAlready {
+				continue
+			}
+
+			elements = append(elements, struct {
+				elem      interface{}
+				elemBytes []byte
+			}{elem, elemBytes})
+			elementsMap[elemBytesString] = customtypes.Void
+		}
+		if fieldMetadata.LexicalOrder {
```

Comment:
Check the lexical order up on each iteration instead of after having read already all elements, as you know during the iteration already, whether the ordering is violated by comparing to just the previous element.

Reply:
The question is what behavior to expect when the order is violated - currently it deserializes whatever order is provided, and then sorts the elements. Should it return an error instead?